designs/generic-git-provider-gitops.md

```diff
@@ -0,0 +1,205 @@
+
+The goal of this design is to unlock the ability for EKS-A users to use any git provider which supports generic authentication methods like SSH authentication as their GitOps provider.
+
+## Tenants
+- Supports any generic generic git providers with password/SSH auth
+- Minimal -- simple and straight forward
+- No tight coupling preventing abstraction of Flux configuration into a curated package in the future
+
+## Scope
+### In-Scope
+- addition of a generic git provider
+- SSH and password auth for generic git provider
+
+### Out-of-Scope
+- provider-specific actions (create, delete repository) for aribitrary git providers
+- refactoring to use git executable vs gogit library
+- refactoring to use flux git provider Go libraries
+
+## Overview of Solution
+Currently, EKS Anywhere uses a [Github specific Flux bootstrap](https://fluxcd.io/docs/cmd/flux_bootstrap_github/) command to install and bootstrap flux on a new EKS Anywhere cluster (`flux bootstrap github`) and uses the Github API to manage the repository itself (create it during `create cluster`, delete it during `delete cluster`).
+
+In order to support the broadest array of git hosting solutions via basic authentication like SSH and password/username, the generic git provider will simplify this process, accepting basic authentication parameters and passing them to Flux's [`flux bootstrap git`](https://fluxcd.io/docs/cmd/flux_bootstrap_git/) command, bypassing the repository management steps we would take for a specific provider. This will enable users who rely on a variety of Git providers to use GitOps with EKS Anywhere right away.
+
+## User Experience
+
+### Create
+The EKS Anywhere CLI user will provide the configuration neccessary to authenticate to their generic git repository via a `git` provider in the `gitOpsConfig` spec. The changes to this spec are discussed in more detail below.
+
+```yaml
+apiVersion: anywhere.eks.amazonaws.com/v1alpha1
+kind: GitOpsConfig
+metadata:
+  name: my-generic-git-provider-gitops
+spec:
+  flux:
+    fluxSystemNamespace: "some-other-namespace"
+    git:
+      repositoryUrl: git@github.com:danbudris/design-doc-cluster.git
+      username: danbudris
+      privateKeyFile: '/path/to/my/private_key.rsa'
+      branch: branchToUseInRemoteRepo
+```
+
+The CLI will then validate if the repository exists; if it does not exist, it will return a preflight validation error, as we cannot create a repository in a remote with basic auth credentials and no access to the API.
+
+If it does exist, the installation will proceed in the same manner as the current Github integration, cloning the repository locally, initializing it if needed, and committing the EKS Anywhere/Flux system configuration files and pushing the contents to the remote.
+
+### Delete
+The delete experience will bypass the deletion of the repository if the generic git provider is used.
```

Comment:
What about deleting the files from the respository?

Reply:
yes, this will be covered at delete time, as we can use basic git actions and filesystem actions (fetch, clone, rm, etc) to do this.